File: src/main/java/pe/nanamochi/banchus/entities/packets/Match.java
package pe.nanamochi.banchus.entities.packets;

import java.util.List;
import lombok.AllArgsConstructor;
import lombok.Data;
import lombok.NoArgsConstructor;
import pe.nanamochi.banchus.entities.commons.MatchTeamType;
import pe.nanamochi.banchus.entities.commons.MatchType;
import pe.nanamochi.banchus.entities.commons.Mode;
import pe.nanamochi.banchus.entities.commons.ScoringType;

@Data
@NoArgsConstructor
@AllArgsConstructor
public class Match {
  private int id;
  private boolean inProgress;
  private MatchType type;
  private int mods;
  private String name;
  private String password;
  private String beatmapName;
  private int beatmapId;
  private String beatmapMd5;
  private List<MatchSlot> slots;
  private int hostId;
  private Mode mode;
  private ScoringType scoringType;
  private MatchTeamType teamType;
  private boolean freemodsEnabled;
  private int randomSeed;
}

File: banchus-app/src/main/kotlin/pe/nanamochi/banchus/util/Match.kt
package pe.nanamochi.banchus.util

import pe.nanamochi.banchus.components.MatchSlot
import pe.nanamochi.banchus.components.MatchTeamType
import pe.nanamochi.banchus.components.Mode
import pe.nanamochi.banchus.components.ScoringType
import pe.nanamochi.banchus.components.SlotTeam
import pe.nanamochi.banchus.redis.entity.MultiplayerMatch
import pe.nanamochi.banchus.redis.entity.MultiplayerMatchSlot

fun MultiplayerMatch.asBancho(slots: List<MultiplayerMatchSlot>) =
    pe.nanamochi.banchus.components.Match(
        id = inGameMatchId().toInt(),
        inProgress = inProgress,
        mods = mods,
        name = name,
        password = password,
        beatmapName = beatmapName,
        beatmapMd5 = beatmapMd5,
        beatmapId = beatmapId,
        slots = slots.asBancho(),
        hostId = hostUserId,
        mode = Mode.fromValue(mode),
        scoringType = ScoringType.fromValue(winCondition.toInt()),
        teamType = MatchTeamType.fromValue(teamType.toInt()),
        freemodsEnabled = freemodEnabled,
        randomSeed = randomSeed.toUInt(),
    )

fun List<MultiplayerMatchSlot>.asBancho() =
    this.map { slot ->
        MatchSlot(
            userId = slot.user?.userId ?: 0,
            status = slot.status.value.toByte(),
            team = SlotTeam.fromValue(slot.team.value.toInt()),
            mods = slot.mods,
        )
    }
